### PR TITLE
Major updates to the pulpcore policy

### PR DESCRIPTION
**Pull request by zpytela**
Support for RHEL 8.
Support for ansible.
Different types for client and server domains, together with separate
private types for /run, /tmp, /var/lib.
Support for multiple paths for the same executable.

### CODE REVIEW

**Comment by zpytela** [comment]
Note this repository contains 3 files for the policy to be used right away, without backporting some commits from upstream selinux policy. They should be installed in this order:
```
semodule -i pulpcore_port.pp
semodule -i pulpcore.pp
semodule -i pulpcore_rhsmcertd.pp
semanage port -a -t pulpcore_port_t -p tcp 24816-24817
```
Still, one additional policy is desirable to have so that directories in /run are created with the right context, but it does not currently block testing.

File: pulpcore.fc
Comment (ekohl) [inline pulpcore.fc:1]: Not everything in `/etc/pulp` belongs to Pulp 3. On migrations there's also Pulp 2 content there.

File: pulpcore.fc
Comment (ekohl) [inline pulpcore.fc:21]: I wonder: `/var/run` is now usually a symlink to `/run` so is that really relevant? At least in in https://github.com/theforeman/puppet-pulpcore/pull/116 (the Puppet based installer) I'm dropping the use of these directories altogether by dropping pid files. They're useless on systemd anyway.

File: pulpcore.fc
Comment (ekohl) [inline pulpcore.fc:4]: In https://github.com/theforeman/puppet-pulpcore/pull/116 I don't use these and only use wrappers in `/usr/libexec/pulpcore`. Per https://github.com/pulp/pulpcore-selinux/pull/13 I think that's the only place that should actually have a context. Please consider this.

IMHO we should also modify the Ansible based installer to stop using these and only use libexec wrappers. Then we can drop these 2 and `/usr/local/lib/pulp/bin`.

File: pulpcore.fc
Comment (ekohl) [inline pulpcore.fc:18]: This directory is also used by Pulp 2. I think it may also set a context in `pulp-selinux` which might conflict.

File: pulpcore.te
Comment (ekohl) [inline pulpcore.te:31]: I'd like to remove the use of pid files completely so perhaps we should do that before merging this? (and also get rid of `files_pid_filetrans` below)

File: pulpcore.te
Comment (ekohl) [inline pulpcore.te:87]: I have issues with commented code. Is this needed or can it be removed?

File: pulpcore.te
Comment (ekohl) [inline pulpcore.te:121]: Also commented code

File: pulpcore.fc
Comment (zpytela) [reply pulpcore.fc:1]: If we knew what kind of files can be there, we could assign different types. With this line, all files have the same default private type `pulpcore_etc_t`. Is this a problem?

File: pulpcore.fc
Comment (zpytela) [reply pulpcore.fc:21]: @ekohl, I don't understand. It's a true that /var/run is a symlink to /run, but we have equivalency rules for that. Systemd uses the runtime dirs as well as the actual services on the system I have access to.

File: pulpcore.fc
Comment (zpytela) [reply pulpcore.fc:18]: In my instance, there is also `devel` and `sign-metadata.sh`. I can add them to the enumeration list and remove the regexp line.

File: pulpcore.te
Comment (zpytela) [reply pulpcore.te:87]: A result of testing, no need for commented code.

File: pulpcore.fc
Comment (zpytela) [reply pulpcore.fc:4]: In my instance there are no libexec wrappers.

File: pulpcore.te
Comment (zpytela) [reply pulpcore.te:31]: If the service runtime content is at a different place, we can remove it. Just note with a generic `var_run_t` type for the pidfile the service will not have access to it.

File: pulpcore.fc
Comment (mikedep333) [reply pulpcore.fc:4]: @ekohl @zpytela we should keep the `/usr/local/lib/pulp/bin/` filepaths, per the previous convo in a PR (which I finally replied to)
https://github.com/pulp/pulpcore-selinux/pull/13#issuecomment-693544960

**Comment by mikedep333** [comment]
> Note this repository contains 3 files for the policy to be used right away, without backporting some commits from upstream selinux policy. They should be installed in this order:
> 
> ```
> semodule -i pulpcore_port.pp
> semodule -i pulpcore.pp
> semodule -i pulpcore_rhsmcertd.pp
> semanage port -a -t pulpcore_port_t -p tcp 24816-24817
> ```

Thank you, I was wondering about pulpcore_port in particular.

The Makefile is different on the development node. Is that something you meant to update?

Also, we have a README.md in this repo.

**Comment by mikedep333** [comment]
We previously discussed having a boolean for ansible plugin support or not (late in development), but I do not see one here. Ansible plugin support is baked in.

@bmbouter Are you OK with there not being such a boolean?

**Comment by zpytela** [comment]
> > Note this repository contains 3 files for the policy to be used right away, without backporting some commits from upstream selinux policy. They should be installed in this order:
> > ```
> > semodule -i pulpcore_port.pp
> > semodule -i pulpcore.pp
> > semodule -i pulpcore_rhsmcertd.pp
> > semanage port -a -t pulpcore_port_t -p tcp 24816-24817
> > ```
> 
> Thank you, I was wondering about pulpcore_port in particular.
> 
> The Makefile is different on the development node. Is that something you meant to update?
I haven't changed Makefile in my PR.

> 
> Also, we have a README.md in this repo.
Do you mean I should update it with the instructions?

**Comment by mikedep333** [comment]
> > > Note this repository contains 3 files for the policy to be used right away, without backporting some commits from upstream selinux policy. They should be installed in this order:
> > > ```
> > > semodule -i pulpcore_port.pp
> > > semodule -i pulpcore.pp
> > > semodule -i pulpcore_rhsmcertd.pp
> > > semanage port -a -t pulpcore_port_t -p tcp 24816-24817
> > > ```
> > 
> > 
> > Thank you, I was wondering about pulpcore_port in particular.
> > The Makefile is different on the development node. Is that something you meant to update?
> I haven't changed Makefile in my PR.

Great, I won't worry about it.

> > Also, we have a README.md in this repo.
> Do you mean I should update it with the instructions?

You should. I won't block the merge on it, since it seems very straightforward.

**Comment by bmbouter** [comment]
> We previously discussed having a boolean for ansible plugin support or not (late in development), but I do not see one here. Ansible plugin support is baked in.
> 
> @bmbouter Are you OK with there not being such a boolean?

Yes I am. We want to launch a working policy, we can create a more fine-grained policy as a refinement.

**Comment by zpytela** [comment]
> > We previously discussed having a boolean for ansible plugin support or not (late in development), but I do not see one here. Ansible plugin support is baked in.
> > @bmbouter Are you OK with there not being such a boolean?
> 
> Yes I am. We want to launch a working policy, we can create a more fine-grained policy as a refinement.
After adding the ansible test to the test suite, these permissions were added:
- /dev/shm handling
- name_connect on 24817/tcp
If you think it's worth a boolean, it can be changed. Without further understanding I cannot assess if it is worth the change. It may also depend if we decide to make some major changes in the current policy draft.

**Comment by bmbouter** [comment]
> > > We previously discussed having a boolean for ansible plugin support or not (late in development), but I do not see one here. Ansible plugin support is baked in.
> > > @bmbouter Are you OK with there not being such a boolean?
> > 
> > 
> > Yes I am. We want to launch a working policy, we can create a more fine-grained policy as a refinement.
> > After adding the ansible test to the test suite, these permissions were added:
> 
>     * /dev/shm handling
> 
>     * name_connect on 24817/tcp
>       If you think it's worth a boolean, it can be changed. Without further understanding I cannot assess if it is worth the change. It may also depend if we decide to make some major changes in the current policy draft.

If you could group these into a section and put a comment that these X statements are `pulp_ansible` related I think that would be all we need to do for now.

**Comment by zpytela** [comment]
> > > > We previously discussed having a boolean for ansible plugin support or not (late in development), but I do not see one here. Ansible plugin support is baked in.
> > > > @bmbouter Are you OK with there not being such a boolean?
> > > 
> > > 
> > > Yes I am. We want to launch a working policy, we can create a more fine-grained policy as a refinement.
> > > After adding the ansible test to the test suite, these permissions were added:
> > 
> > 
> > ```
> > * /dev/shm handling
> > 
> > * name_connect on 24817/tcp
> >   If you think it's worth a boolean, it can be changed. Without further understanding I cannot assess if it is worth the change. It may also depend if we decide to make some major changes in the current policy draft.
> > ```
> 
> If you could group these into a section and put a comment that these X statements are `pulp_ansible` related I think that would be all we need to do for now.
Grouping is not easy to achieve when following the structure; furthermore, I can only say these permissions were added *after* running the tests. Note I am not so familiar with how things inside pulp3 actually work. Additionally, it is possible some permissions will be found necessary to allow later.

For now, I can split the PR into 2 commits; however, as the policy can change any time later, I can't see a direct benefit from that.

**Comment by zpytela** [comment]
> > > > Note this repository contains 3 files for the policy to be used right away, without backporting some commits from upstream selinux policy. They should be installed in this order:
> > > > ```
> > > > semodule -i pulpcore_port.pp
> > > > semodule -i pulpcore.pp
> > > > semodule -i pulpcore_rhsmcertd.pp
> > > > semanage port -a -t pulpcore_port_t -p tcp 24816-24817
> > > > ```
> > > 
> > > 
> > > Thank you, I was wondering about pulpcore_port in particular.
> > > The Makefile is different on the development node. Is that something you meant to update?
> > > I haven't changed Makefile in my PR.
> 
> Great, I won't worry about it.
> 
> > > Also, we have a README.md in this repo.
> > > Do you mean I should update it with the instructions?
> 
> You should. I won't block the merge on it, since it seems very straightforward.
Updated.

> 
> Please work with @ekohl on his comments.
I addressed the comments I can confirm. I cannot address the comments left unaddressed so far until I can see the changed environment or any other resource which can be used for the policy adjustment. The current policy reflects the state in the instance I have access to.

File: pulpcore.fc
Comment (ekohl) [reply pulpcore.fc:1]: I can't find the git repo for the old pulp-selinux package, but what I see today:
```console
# ls -lZ /etc/pulp
drwxr-xr-x. root   root   system_u:object_r:httpd_sys_content_t:s0 content
-rw-r--r--. root   root   system_u:object_r:httpd_sys_content_t:s0 repo_auth.conf
drwxr-xr-x. root   root   system_u:object_r:httpd_sys_content_t:s0 server
-rw-------. apache apache system_u:object_r:httpd_sys_content_t:s0 server.conf
-rw-r-----. root   pulp   system_u:object_r:httpd_sys_content_t:s0 settings.py
-rw-r--r--. root   root   system_u:object_r:httpd_sys_content_t:s0 streamer.conf
drwxr-xr-x. root   root   system_u:object_r:httpd_sys_content_t:s0 vhosts80
```

This comes from:
```console
# semanage fcontext --list | grep pulp
/etc/pulp(/.*)?                                    all files          system_u:object_r:httpd_sys_content_t:s0 
/var/run/pulp(/.*)?                                all files          system_u:object_r:pulp_var_run_t:s0 
/etc/pki/pulp(/.*)?                                all files          system_u:object_r:pulp_cert_t:s0 
/var/lib/pulp(/.*)?                                all files          system_u:object_r:httpd_sys_rw_content_t:s0 
/var/log/pulp(/.*)?                                all files          system_u:object_r:httpd_sys_rw_content_t:s0 
/var/cache/pulp(/.*)?                              all files          system_u:object_r:pulp_var_cache_t:s0 
/var/lib/pulp/tmp(/.*)?                            all files          system_u:object_r:pulpcore_var_lib_t:s0 
/var/run/pulpcore*(/.*)?                           all files          system_u:object_r:pulpcore_var_run_t:s0 
/usr/share/pulp/wsgi(/.*)?                         all files          system_u:object_r:httpd_sys_content_t:s0 
/var/lib/pulp/assets(/.*)?                         all files          system_u:object_r:pulpcore_var_lib_t:s0 
/var/lib/pulp/upload(/.*)?                         all files          system_u:object_r:pulpcore_var_lib_t:s0 
/var/run/pulpcore-api(/.*)?                        all files          system_u:object_r:pulpcore_var_run_t:s0 
/var/lib/pulp/artifact(/.*)?                       all files          system_u:object_r:pulpcore_var_lib_t:s0 
/var/run/pulpcore-content(/.*)?                    all files          system_u:object_r:pulpcore_var_run_t:s0 
/var/run/pulpcore-resource-worker[0-9]*(/.*)?      all files          system_u:object_r:pulpcore_var_run_t:s0 
/var/run/pulpcore-resource-manager(/.*)?           all files          system_u:object_r:pulpcore_var_run_t:s0 
/usr/lib/pulp/bin/rq                               regular file       system_u:object_r:pulpcore_exec_t:s0 
/usr/bin/pulp_streamer                             regular file       system_u:object_r:streamer_exec_t:s0 
/usr/lib/pulp/bin/gunicorn                         regular file       system_u:object_r:pulpcore_exec_t:s0 
/usr/local/lib/pulp/bin/rq                         regular file       system_u:object_r:pulpcore_exec_t:s0 
/usr/local/lib/pulp/bin/gunicorn                   regular file       system_u:object_r:pulpcore_exec_t:s0 
```

Note the first line - this is an exact duplicate of that. In Satellite the migration plan is to install both in parallel and online migrate from 2 to 3. So both must be able to operate in parallel.

You're more familiar with SELinux so I hope you can tell me what SELinux does if there are 2 patterns that are exactly the same, but I can't imagine that works well.

File: pulpcore.fc
Comment (ekohl) [reply pulpcore.fc:22]: To be clear: I was suggesting that the systemd unit files should be modified to not use pid files anymore, removing the need for these run dirs. That's probably more something for the installer maintainers

File: pulpcore.fc
Comment (zpytela) [reply pulpcore.fc:1]: For each path there can be only one rule, just more specific rules have priority over regexps.
I can drop the pulpcore_etc_t support completely, but on the system I have access to there is no default rule so the files will get etc_t.

File: pulpcore.fc
Comment (zpytela) [reply pulpcore.fc:22]: In my instance there are not only pid files, but also other files in /run/pulpcore-*/, so dropping the support right away means the tests will stop working. I don't have any other way how to verify the policy is sufficient.

File: pulpcore.fc
Comment (ekohl) [reply pulpcore.fc]: IMHO the cleanest solution is to use `/etc/pulpcore` and place `settings.py` there, but that's something that needs more planning. Should we make it more specific and explicitly list `/etc/pulpcore/settings.py`? AFAIK that's the only file in it actually needs. That probably warrants a comment, something like `# More specific to avoid conflicts with Pulp 2's SELinux policy`.

Comment (zpytela) [reply pulpcore.fc]: From SELinux POV no need for a different directory, explicit path would work. There are also 2 keys there.

File: pulpcore.fc
Comment (ekohl) [reply pulpcore.fc:20]: I'd suggest to keep it in the SELinux policy, but I'd ask @mikedep333 to move away from pid files in the Ansible installer so this is no longer needed.

Comment (zpytela) [reply pulpcore.fc:20]: I am also for keeping it now, note there are a dozen of lines related to runtime files usage. Once there are some changes in the code or installer, the policy would need to be adjusted anyway, so we can drop these lines.

**Review by mikedep333 — APPROVED** [review]

**Comment by mikedep333** [comment]
@ekohl I think we should merge now, and address any issues after actually testing them. @Spredzy is working on pulp_installer's ability to install the policy, and you can package and test them for Katello.

@bmbouter Please merge, since I cannot.

**Comment by bmbouter** [comment]
I agree let's merge. I want to hear one last ack form @ekohl first. @ekohl is this lgtm? @mikedep333 I will merge then, np.

**Comment by mikedep333** [comment]
I think we should merge now actually. My reasoning is that there are guaranteed to be some more differences between the old pulp_installer environment (EC2 instance) that I created & @zpytela used, and the current pulp_installer environment, so we should address them after merge. @Spredzy is working on pulp_installer's ability to install the policy.

While we do this, the Katello team can package pulpcore-selinux and address the differences for Katello

**Comment by bmbouter** [comment]
@mikedep333 that makes perfect sense to me. We really need this merged and testing in the various places asap. I appreciate you helping me get the timeline right.

@ekohl if there is more to be done post-merge, please let us know! Thanks for helping us test this.

@zpytela thank you soooooo much! :rocket: :100:

**Pull request merged**
